staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/listtype/validation.go

```diff
@@ -0,0 +1,225 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package listtype
+
+import (
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+)
+
+// ValidateListSetsAndMaps validates that arrays with x-kubernetes-list-type "map" and "set" fulfill the uniqueness
+// invariants for the keys (maps) and whole elements (sets).
+func ValidateListSetsAndMaps(fldPath *field.Path, s *schema.Structural, obj map[string]interface{}) field.ErrorList {
+	if s == nil || obj == nil {
+		return nil
+	}
+
+	var errs field.ErrorList
+
+	if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+		for k, v := range obj {
+			errs = append(errs, validationListSetAndMaps(fldPath.Key(k), s.AdditionalProperties.Structural, v)...)
+		}
+	}
+	if s.Properties != nil {
+		for k, v := range obj {
+			if sub, ok := s.Properties[k]; ok {
+				errs = append(errs, validationListSetAndMaps(fldPath.Child(k), &sub, v)...)
+			}
+		}
+	}
+
+	return errs
+}
+
+func validationListSetAndMaps(fldPath *field.Path, s *schema.Structural, obj interface{}) field.ErrorList {
+	switch obj := obj.(type) {
+	case []interface{}:
+		return validateListSetsAndMapsArray(fldPath, s, obj)
+	case map[string]interface{}:
+		return ValidateListSetsAndMaps(fldPath, s, obj)
+	}
+	return nil
+}
+
+func validateListSetsAndMapsArray(fldPath *field.Path, s *schema.Structural, obj []interface{}) field.ErrorList {
+	var errs field.ErrorList
+
+	if s.XListType != nil {
+		switch *s.XListType {
+		case "set":
+			nonUnique, err := validateListSet(fldPath, obj)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				for _, i := range nonUnique {
+					errs = append(errs, field.Duplicate(fldPath.Index(i), obj[i]))
+				}
+			}
+		case "map":
+			errs = append(errs, validateListMap(fldPath, s, obj)...)
+		}
+	}
+
+	if s.Items != nil {
+		for i := range obj {
+			errs = append(errs, validationListSetAndMaps(fldPath.Index(i), s.Items, obj[i])...)
+		}
+	}
+
+	return errs
+}
+
+// validateListSet validated uniqueness of unstructured objects (scalar and compound) and
+// returns the first non-unique appearance of items.
+//
+// As a special case to distinguish undefined key and null values, we allow unspecifiedKeyValue and nullObjectValue
+// which are both handled like scalars with correct comparison by Golang.
+func validateListSet(fldPath *field.Path, obj []interface{}) ([]int, *field.Error) {
+	if len(obj) <= 1 {
+		return nil, nil
+	}
+	if len(obj) == 2 {
+		if reflect.DeepEqual(obj[0], obj[1]) {
```

Comment:
are there cases where this would return false, but json-marshalling and comparing would be identical?

Reply:
`json.Marshal` is the inverse of `json.Unmarshal`. If both values are the result of `json.Unmarshal` then this will not happen.

What do you have in mind? Our patch logic?

Reply:
more issues like https://github.com/json-iterator/go/issues/323, where json-iterator normalizes on marshal as opposed to encoding/json normalizing on both unmarshal/marshal

Reply:
ic. Suggestion? Marshal again?

Reply:
that is definitely simpler to reason about; that's what I would do to start